# 异常处理器

每层调用，如果有异常了，有的可能是直接代码内部 try catch，也有可能是向上抛，从 service 层抛到 controller 层，controller 再抛到前端控制器，前端控制器再抛就抛出后台了，被用户看到了，用户肯定是不懂代码的，让前端去处理也是不可能的，所以后端必须处理掉异常

前端控制器有一个组件，==HandlerExceptionResolver 异常处理器== 通过该组件来处理全局异常



![image-20210720152827212](https://attach.blog.wen7.online/image-20210720152827212.png)



## 设置全局异常处理器

步骤：

1. 创建一个全局异常处理器的类
2. 加上注解 @ControllerAdvice
3. 编写方法，加上注解 @ResponseBody 和 @ExceptionHandler(value = {Exception.class})
5. 编写自定义的异常处理逻辑

```java
package com.example.demo.config;

import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * 全局异常处理器
 */
@ControllerAdvice
public class GlobalExceptionHandler{

    @ResponseBody
    @ExceptionHandler(value = {Exception.class})
    public String resolveException(HttpServletRequest request, HttpServletResponse response, Exception e) {
        /**
         * 具体的业务逻辑
         */
        return "自定义返回信息";
    }
}
```

==@ControllerAdvice==  详情请参考独立文档



## @ExceptionHandler(value = {Exception.class})

可以指定多个异常，一般情况指定一个 Exception 可以捕获所有的异常，在方法内部进行不同异常判断及处理

如果如下代码，多个异常处理，如果有 ArrayIndexOutOfBoundsException 异常会优先进入方法 resolveArrayIndexOutOfBoundsException

```java
@ControllerAdvice
public class GlobalExceptionHandler{

    @ResponseBody
    @ExceptionHandler(Exception.class)
    public String resolveException(HttpServletRequest request, HttpServletResponse response, Exception e) {
        // 具体的业务逻辑
        return "自定义返回信息";
    }
    
    @ResponseBody
    @ExceptionHandler(ArrayIndexOutOfBoundsException.class)
    public String resolveArrayIndexOutOfBoundsException(HttpServletRequest request, HttpServletResponse response, Exception e) {
        // 具体的业务逻辑
        return "自定义返回信息";
    }
}
```

